### PR TITLE
*: fix issue 1545 and 1536

### PR DESCRIPTION
**Pull request by tiancaiamao**
ref https://github.com/pingcap/tidb/issues/1545  https://github.com/pingcap/tidb/issues/1536

### CODE REVIEW

File: plan/resolver.go [hunk 1]
```diff
@@ -953,3 +964,29 @@ func (nr *nameResolver) fillShowFields(s *ast.ShowStmt) {
 	s.SetResultFields(fields)
 	nr.currentContext().fieldList = fields
 }
+
+func (nr *nameResolver) checkCreateTableGrammar(stmt *ast.CreateTableStmt) {
```
Comment (shenli) [inline plan/resolver.go]: It is better to move this function to validator.go

File: plan/validator.go [hunk 2]
```diff
@@ -256,3 +267,29 @@ func (v *validator) checkFieldList(x *ast.FieldList) {
...
+				return
+			}
+		}
+	}
+}
+
+func (v *validator) checkCreateIndexGrammar(stmt *ast.CreateIndexStmt) {
+	for i := 0; i < len(stmt.IndexColNames); i++ {
+		name1 := stmt.IndexColNames[i].Column.Name
+		for j := 0; j < len(stmt.IndexColNames); j++ {
```
Comment (shenli) [inline plan/validator.go]: j := i+1 or use a map to avoid n*n complexity.

**Comment by shenli** [comment]
PTAL @coocood @zimulala

File: plan/validator.go (hunk 2 above)
Comment (tiancaiamao) [inline plan/validator.go]: is that necessary?  
I guess columns should be rather small, say, n<5 or n<10, then we don't need care the O(n^2)  avoid some allocations is also a bonus.

Comment (shenli) [inline plan/validator.go]: At least you should change j:= 0 to j:=i+1.

**Comment by shenli** [comment]
LGTM

File: plan/validator.go [hunk 3]
```diff
@@ -256,3 +267,27 @@ func (v *validator) checkFieldList(x *ast.FieldList) {
 		}
 	}
 }
+
+func (v *validator) checkCreateTableGrammar(stmt *ast.CreateTableStmt) {
+	for _, colDef := range stmt.Cols {
+		if colDef.Tp.Tp == mysql.TypeString {
+			if colDef.Tp.Flen != types.UnspecifiedLength && colDef.Tp.Flen > 255 {
```
Comment (zimulala) [inline plan/validator.go]: Why not put these conditions in a ‘if’ statement ？

**Comment by zimulala** [comment]
LGTM

**Pull request merged**